nixos/modules/services/networking/netbird.nix

```diff
@@ -1,171 +1,418 @@
                   {
-                    NB_CONFIG = "/var/lib/''${stateDir}/config.json";
-                    NB_LOG_FILE = "console";
-                    NB_WIREGUARD_PORT = builtins.toString port;
-                    NB_INTERFACE_NAME = name;
-                    NB_DAMEON_ADDR = "/var/run/''${stateDir}"
+                    NB_CONFIG = "/var/lib/''${config.name}/config.json";
+                    NB_DAEMON_ADDR = "unix:///var/run/''${config.name}/sock";
+                    NB_INTERFACE_NAME = config.interface;
+                    NB_LOG_FILE = mkOptionDefault "console";
+                    NB_LOG_LEVEL = config.logLevel;
+                    NB_WIREGUARD_PORT = toString config.port;
                   }
                 '';
                 description = ''
                   Environment for the netbird service, used to pass configuration options.
                 '';
               };
 
-              stateDir = mkOption {
-                type = str;
-                default = "netbird-${name}";
-                description = ''
-                  Directory storing the netbird configuration.
-                '';
+              autoStart = mkOption {
+                type = bool;
+                default = true;
+                description = "start service with the system";
+              };
+
+              hardened = mkOption {
+                type = bool;
+                default = true;
+                description = "harden the service by running as ephemeral user with minimal set of permissions";
+              };
+
+              logLevel = mkOption {
+                type = enum [
+                  # logrus loglevels
+                  "panic"
+                  "fatal"
+                  "error"
+                  "warn"
+                  "warning"
+                  "info"
+                  "debug"
+                  "trace"
+                ];
+                default = "info";
+                description = "log level of this client";
+              };
+
+              wrapper = mkOption {
+                type = package;
+                internal = true;
+                default =
+                  let
+                    exports = pipe config.environment [
+                      (mapAttrsToList (name: default:
+                        # Couldn't default with `''${${name}:${escapeShellArg default}}`
+                        # because `escapeShellArg` wraps values in single quotes,
+                        # which in turn works differenty (or rather doesn't) than double-quotes
+                        # in this shell parameter expansion
+                        ''test "''${${name}+X}" == X || export ${toShellVar name default}''))
+                      (concatStringsSep "\n")
+                    ];
+                  in
+                  pkgs.writeShellScriptBin "netbird${config.suffix}" ''
+                    ${exports}
+                    ${lib.getExe' cfg.package "netbird"} "$@"
+                  '';
```

Comment:
This is very weird, nixpkgs has a machinery already to wrap programs c.f. https://nixos.org/manual/nixpkgs/stable/#fun-wrapProgram

Reply:
I'm pretty sure it would be significantly longer with `wrapProgram`, something like:
```
mkDerivation {
	pname = "xxx";
    installPhase = 
	    let
	  		exports = pipe config.environment [
	          (mapAttrsToList (name: default:
	            "--set-default '${name}' '${default}'"))
	          (concatStringsSep " ")
	        ];
		in
		"wrapProgram ${lib.getExe' cfg.package "netbird"} $out/bin/netbird${config.suffix} ${exports}";
}
```

actually makes more sense after prototyping it and seeing with own eyes

Reply:
You can do something like
```nix
    stdenv.mkDerivation {
      intallPhase = ''
        wrapProgram ${lib.getExe cfg.package} $out/bin/netbird${config.suffix} ${
          concatMapStringsSep " " ({ name, value }: "--set-default '${name}' '${default}'") (
            attrsToList config.environment
          )
        }
      '';
```
The issue I had was not because of the size of the code, but reinventing the wheel ^^

Reply:
not so short anymore:
```nix
let
  makeWrapperArgs = flatten (mapAttrsToList
    (key: value: [ "--set-default" key value ])
    config.environment
  );
in
pkgs.stdenv.mkDerivation {
  name = "${cfg.package.name}-wrapper-${name}";
  meta.mainProgram = config.name;
  nativeBuildInputs = with pkgs; [ makeWrapper ];
  phases = [ "installPhase" ];
  installPhase = ''
    mkdir -p "$out/bin"
    makeWrapper ${lib.getExe cfg.package} "$out/bin/${config.name}" \
      ${escapeShellArgs makeWrapperArgs}
  '';
}
```